README.md
# nvm-windows

[![Build status](https://ci.appveyor.com/api/projects/status/v0cexq0t8yntw6o7/branch/cn?svg=true)](https://ci.appveyor.com/project/alvis-zhao/nvm-windows/branch/cn)

个人定制版, 修复官方一些已知的问题

## 编译测试环境

1. go 1.16 windows/amd64

## 更新记录

见 [更新文档](./CHANGELOG.md)

## 注意事项

* 编译时设置go代理方便下载依赖[代理设置](https://goproxy.io/zh/docs/getting-started.html)
* 必须使用管理员权限运行本工具


CHANGELOG.md
# 更新日志

## 1.1.9

* 默认使用淘宝源，避免每次都要设置代理

## 1.1.8

* 修复路径存在空格导致无法编译的问题[#588](https://github.com/coreybutler/nvm-windows/issues/588)
* 修复windows 10 64 位下无法编译的问题[#548](https://github.com/coreybutler/nvm-windows/issues/548)
* 默认构建为64位(GOARCH=amd64)
* 修复go 1.16 利用go module构建的问题, 并启用appveyor ci


src/node/node.go
package node

import (
	"encoding/json"
	"io/ioutil"
	"os/exec"
	"regexp"
	"strings"

	"alvisisme/nvm/src/arch"
	"alvisisme/nvm/src/file"
	"alvisisme/nvm/src/web"

	"github.com/blang/semver/v4"
)

/**
 * Returns version, architecture
 */
func GetCurrentVersion() (string, string) {
	cmd := exec.Command("node", "-v")
	str, err := cmd.Output()
	if err == nil {
		v := strings.Trim(regexp.MustCompile("-.*$").ReplaceAllString(regexp.MustCompile("v").ReplaceAllString(strings.Trim(string(str), " \n\r"), ""), ""), " \n\r")
		cmd := exec.Command("node", "-p", "console.log(process.execPath)")
		str, _ := cmd.Output()
		file := strings.Trim(regexp.MustCompile("undefined").ReplaceAllString(string(str), ""), " \n\r")
		bit := arch.Bit(file)
		if bit == "?" {
			cmd := exec.Command("node", "-e", "console.log(process.arch)")
			str, err := cmd.Output()
			if err == nil {
				if string(str) == "x64" {
					bit = "64"
				} else {
					bit = "32"
				}
			} else {
				return v, "Unknown"
			}
		}
		return v, bit
	}
	return "Unknown", ""
}

func IsVersionInstalled(root string, version string, cpu string) bool {
	e32 := file.Exists(root + "\\v" + version + "\\node32.exe")
	e64 := file.Exists(root + "\\v" + version + "\\node64.exe")
	used := file.Exists(root + "\\v" + version + "\\node.exe")
	if cpu == "all" {
		return ((e32 || e64) && used) || e32 && e64
	}
	if file.Exists(root + "\\v" + version + "\\node" + cpu + ".exe") {
		return true
	}
	if ((e32 || e64) && used) || (e32 && e64) {
		return true
	}
	if !e32 && !e64 && used && arch.Validate(cpu) == arch.Bit(root+"\\v"+version+"\\node.exe") {
		return true
	}
	if cpu == "32" {
		return e32
	}
	if cpu == "64" {
		return e64
	}
	return false
}

func IsVersionAvailable(v string) bool {
	// Check the service to make sure the version is available
	avail, _, _, _, _, _ := GetAvailable()

	for _, b := range avail {
		if b == v {
			return true
		}
	}
	return false
}

func reverseStringArray(str []string) []string {
	for i := 0; i < len(str)/2; i++ {
		j := len(str) - i - 1
		str[i], str[j] = str[j], str[i]
	}

	return str
}

func GetInstalled(root string) []string {
	list := make([]semver.Version, 0)
	files, _ := ioutil.ReadDir(root)

	for i := len(files) - 1; i >= 0; i-- {
		if files[i].IsDir() {
			isnode, _ := regexp.MatchString("v", files[i].Name())

			if isnode {
				currentVersionString := strings.Replace(files[i].Name(), "v", "", 1)
				currentVersion, _ := semver.Make(currentVersionString)

				list = append(list, currentVersion)
			}
		}
	}

	semver.Sort(list)

	loggableList := make([]string, 0)

	for _, version := range list {
		loggableList = append(loggableList, "v"+version.String())
	}

	loggableList = reverseStringArray(loggableList)

	return loggableList
}

// Sorting
type BySemanticVersion []string

func (s BySemanticVersion) Len() int {
	return len(s)
}

func (s BySemanticVersion) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

func (s BySemanticVersion) Less(i, j int) bool {
	v1, _ := semver.Make(s[i])
	v2, _ := semver.Make(s[j])
	return v1.GTE(v2)
}

// Identifies a version as "LTS"
func isLTS(element map[string]interface{}) bool {
	switch datatype := element["lts"].(type) {
	case bool:
		return datatype
	case string:
		return true
	}
	return false
}

// Identifies a version as "current"
func isCurrent(element map[string]interface{}) bool {
	if isLTS(element) {
		return false
	}

	version, _ := semver.Make(element["version"].(string)[1:])
	benchmark, _ := semver.Make("1.0.0")

	if version.LT(benchmark) {
		return false
	}

	return true
	// return version.Major%2 == 1
}

// Identifies a stable old version.
func isStable(element map[string]interface{}) bool {
	if isCurrent(element) {
		return false
	}

	version, _ := semver.Make(element["version"].(string)[1:])

	if version.Major != 0 {
		return false
	}

	return version.Minor%2 == 0
}

// Identifies an unstable old version.
func isUnstable(element map[string]interface{}) bool {
	if isStable(element) {
		return false
	}

	version, _ := semver.Make(element["version"].(string)[1:])

	if version.Major != 0 {
		return false
	}

	return version.Minor%2 != 0
}

// Retrieve the remotely available versions
func GetAvailable() ([]string, []string, []string, []string, []string, map[string]string) {
	all := make([]string, 0)
	lts := make([]string, 0)
	current := make([]string, 0)
	stable := make([]string, 0)
	unstable := make([]string, 0)
	npm := make(map[string]string)
	url := web.GetFullNodeUrl("index.json")

	// Check the service to make sure the version is available
	text := web.GetRemoteTextFile(url)

	// Parse
	var data = make([]map[string]interface{}, 0)
	json.Unmarshal([]byte(text), &data)

	for _, element := range data {

		var version = element["version"].(string)[1:]
		all = append(all, version)

		if val, ok := element["npm"].(string); ok {
			npm[version] = val
		}

		if isLTS(element) {
			lts = append(lts, version)
		} else if isCurrent(element) {
			current = append(current, version)
		} else if isStable(element) {
			stable = append(stable, version)
		} else if isUnstable(element) {
			unstable = append(unstable, version)
		}
	}

	return all, lts, current, stable, unstable, npm
}


src/web/web.go
package web

import(
  "fmt"
  "net/http"
  "net/url"
  "os"
  "os/signal"
  "io"
  "io/ioutil"
	"strings"
	"syscall"
  "crypto/tls"
  "strconv"
  "alvisisme/nvm/src/arch"
  "alvisisme/nvm/src/file"
)

var client = &http.Client{}
var nodeBaseAddress = "https://npm.taobao.org/mirrors/node/"
var npmBaseAddress = "https://npm.taobao.org/mirrors/npm/"
// var oldNpmBaseAddress = "https://github.com/npm/npm/archive/"

func SetProxy(p string, verifyssl bool){
  if p != "" && p != "none" {
    proxyUrl, _ := url.Parse(p)
    client = &http.Client{Transport: &http.Transport{Proxy: http.ProxyURL(proxyUrl), TLSClientConfig: &tls.Config{InsecureSkipVerify: verifyssl}}}
  } else {
    client = &http.Client{Transport: &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: verifyssl}}}
  }
}

func SetMirrors(node_mirror string, npm_mirror string){
  if node_mirror != "" && node_mirror != "none"{
    nodeBaseAddress = node_mirror;
    if strings.ToLower(nodeBaseAddress[0:4]) != "http" {
      nodeBaseAddress = "http://"+nodeBaseAddress
    }
  }
  if npm_mirror != "" && npm_mirror != "none"{
    npmBaseAddress = npm_mirror;
    if strings.ToLower(npmBaseAddress[0:4]) != "http" {
      npmBaseAddress = "http://"+npmBaseAddress
    }
  }
}

func GetFullNodeUrl(path string) string{
  return nodeBaseAddress+ path;
}

func  GetFullNpmUrl(path string) string{
  return npmBaseAddress + path;
}

func Download(url string, target string, version string) bool {

  output, err := os.Create(target)
  if err != nil {
    fmt.Println("Error while creating", target, "-", err)
  }
  defer output.Close()

  response, err := client.Get(url)
  if err != nil {
    fmt.Println("Error while downloading", url, "-", err)
  }
  defer response.Body.Close()
  c := make(chan os.Signal, 2)
	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
	go func() {
		<-c
		fmt.Println("Download interrupted.Rolling back...")
		output.Close()
		response.Body.Close()
		var err error
		if strings.Contains(target, "node") {
			err = os.RemoveAll(os.Getenv("NVM_HOME") + "\\v" + version)
		} else {
			err = os.Remove(target)
		}
		if err != nil {
			fmt.Println("Error while rolling back", err)
		}
		os.Exit(1)
	}()
  _, err = io.Copy(output, response.Body)
  if err != nil {
    fmt.Println("Error while downloading", url, "-", err)
  }
  if response.Status[0:3] != "200" {
    fmt.Println("Download failed. Rolling Back.")
    err := os.Remove(target)
    if err != nil {
      fmt.Println("Rollback failed.",err)
    }
    return false
  }

  return true
}

func GetNodeJS(root string, v string, a string) bool {

  a = arch.Validate(a)

  vpre := ""
  vers := strings.Fields(strings.Replace(v,"."," ",-1))
  main, _ := strconv.ParseInt(vers[0],0,0)

  if a == "32" {
    if main > 0 {
      vpre = "win-x86/"
    } else {
      vpre = ""
    }
  } else if a == "64" {
    if main > 0 {
      vpre = "win-x64/"
    } else {
      vpre = "x64/"
    }
  }

  url := getNodeUrl ( v, vpre );

  if url == "" {
    //No url should mean this version/arch isn't available
    fmt.Println("Node.js v"+v+" " + a + "bit isn't available right now.")
  } else {
   fileName := root+"\\v"+v+"\\node"+a+".exe"

    fmt.Println("Downloading node.js version "+v+" ("+a+"-bit)... ")

    if Download(url,fileName,v) {
      fmt.Printf("Complete\n")
      return true
    } else {
      return false
    }
  }
  return false

}

func GetNpm(root string, v string) bool {
  url := GetFullNpmUrl("v"+v+".zip")
  // temp directory to download the .zip file
  tempDir := root+"\\temp"

  // if the temp directory doesn't exist, create it
  if (!file.Exists(tempDir)) {
    fmt.Println("Creating "+tempDir+"\n")
    err := os.Mkdir(tempDir, os.ModePerm)
    if err != nil {
      fmt.Println(err)
      os.Exit(1)
    }
  }
  fileName := tempDir+"\\"+"npm-v"+v+".zip"

  fmt.Printf("Downloading npm version "+v+"... ")
  if Download(url,fileName,v) {
    fmt.Printf("Complete\n")
    return true
  } else {
    return false
  }
}

func GetRemoteTextFile(url string) string {
  response, httperr := client.Get(url)
  if httperr != nil {
    fmt.Println("\nCould not retrieve "+url+".\n\n")
    fmt.Printf("%s", httperr)
    os.Exit(1)
  } else {
    defer response.Body.Close()
    contents, readerr := ioutil.ReadAll(response.Body)
    if readerr != nil {
      fmt.Printf("%s", readerr)
      os.Exit(1)
    }
    return string(contents)
  }
  os.Exit(1)
  return ""
}

func IsNode64bitAvailable(v string) bool {
  if v == "latest" {
    return true
  }

  // Anything below version 8 doesn't have a 64 bit version
  vers := strings.Fields(strings.Replace(v,"."," ",-1))
  main, _ := strconv.ParseInt(vers[0],0,0)
  minor, _ := strconv.ParseInt(vers[1],0,0)
  if main == 0 && minor < 8 {
    return false
  }

  // TODO: fixme. Assume a 64 bit version exists
  return true
}

func getNodeUrl (v string,  vpre string) string {
  //url := "http://nodejs.org/dist/v"+v+"/" + vpre + "/node.exe"
  url := GetFullNodeUrl("v"+v+"/" + vpre + "/node.exe")
  // Check online to see if a 64 bit version exists
  _, err := client.Head( url )
  if err != nil {
    return ""
  }
  return url;
}
